### PR TITLE
feat(ext/kv): connect to remote database

### PR DESCRIPTION
**Pull request by losfair**
This patch adds a `remote` backend for `ext/kv`. This supports connection to Deno Deploy and potentially other services compatible with the KV Connect protocol.

### CODE REVIEW

File: ext/kv/remote.rs [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+  async fn open(
+    &self,
+    _state: Rc<RefCell<OpState>>,
+    path: Option<String>,
+  ) -> Result<Self::DB, AnyError> {
+    let Some(url) = path else {
+      return Err(type_error("Missing database url"));
+    };
+
+    let access_token = std::env::var("DENO_ACCESS_TOKEN")
```
Comment (lucacasonato) [inline ext/kv/remote.rs:81]: Needs a permission check for `DENO_ACCESS_TOKEN`.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -173,6 +173,12 @@ jobs:
         uses: actions/setup-node@v3
         with:
           node-version: 18
+      - name: Install protoc
+        uses: arduino/setup-protoc@v2
+        with:
+          version: '21.12'
+          repo-token: '${{ secrets.GITHUB_TOKEN }}'
+        if: '!(github.event_name == ''pull_request'' && matrix.skip_pr)'
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:181]: Will this require additional install steps for contributors? If so https://github.com/denoland/manual/blob/main/references/contributing/building_from_source.md should be updated

File: ext/kv/README.md [hunk 3]
```diff
@@ -1,3 +1,132 @@
...
-This crate provides a key/value store for Deno.
+This crate provides a key/value store for Deno. For an overview of Deno KV,
+please read the [manual](https://deno.land/manual/runtime/kv).
+
+## Storage Backends
+
+Deno KV has a pluggable storage interface that supports multiple backends:
+
+- SQLite - backed by a local SQLite database. This backend is suitable for
+  development and is the default when running locally.
```
Comment (bartlomieju) [inline ext/kv/README.md:11]: FYI people have been asking about alternative local backends: https://github.com/denoland/deno/discussions/18975 can this be addressed somehow?

File: ext/kv/README.md [hunk 4]
```diff
@@ -1,3 +1,132 @@
...
+
+To connect to a KV Connect service, the user provides an HTTP or HTTPS URL to
+`Deno.openKv`. A background task is then spawned to periodically make HTTP POST
+requests to the provided URL to refresh database metadata.
+
+The HTTP `Authorization` header is included and have the format
+`Bearer <access-token>`. The `<access-token>` is a static token issued by the
+service provider. For Deno Deploy, this is the personal access token generated
+from the dashboard. You can specify the access token with the environment
+variable `DENO_ACCESS_TOKEN`.
```
Comment (bartlomieju) [inline ext/kv/README.md:32]: Can we make this `DENO_KV_ACCESS_TOKEN`? Seems overly broad right now

File: ext/kv/README.md [hunk 5]
```diff
@@ -1,3 +1,132 @@
...
+
+Request body is currently unused. The response is a JSON message that satisfies
+the following [JSON Schema](https://json-schema.org/) definition:
```
Comment (bartlomieju) [inline ext/kv/README.md:35]: Suggest to create a file and put it in `cli/schemas`

**Review by bartlomieju — COMMENTED** [review]
Are there any tests that could be added?

File: ext/kv/README.md (hunk 3 above)
Comment (losfair) [reply ext/kv/README.md:11]: Added a note on additional backends

**Comment by losfair** [comment]
Added notes on protobuf compiler to related docs:

- `deno_install`: https://github.com/denoland/deno_install/pull/269
- manual: https://github.com/denoland/manual/pull/717

File: test_util/src/lib.rs [hunk 6]
```diff
@@ -72,6 +80,9 @@ const PORT: u16 = 4545;
 const TEST_AUTH_TOKEN: &str = "abcdef123456789";
 const TEST_BASIC_AUTH_USERNAME: &str = "testuser123";
 const TEST_BASIC_AUTH_PASSWORD: &str = "testpassabc";
+const KV_DATABASE_ID: &str = "55bd5c48-58c2-4eb4-8adb-995c0763622a";
+const KV_ACCESS_TOKEN: &str = "578528e20364923229b08f94386b236f";
+const KV_DATABASE_TOKEN: &str = "iFOw5tcudKE2Cy4MKNuWZjxGkg4";
```
Comment (ry) [inline test_util/src/lib.rs:85]: I assume these are just mock values not real tokens? Can you make them more obviously such - for example:
```
const KV_DATABASE_ID: &str = "11111111-1111-1111-1111-111111111111";
const KV_ACCESS_TOKEN: &str = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
const KV_DATABASE_TOKEN: &str = "MOCKMOCKMOCKMOCKMOCKMOCKMOCK";
```

Comment (lucacasonato) [reply test_util/src/lib.rs:85]: Can you add a test where the first endpoint returns the wrong JSON, and one where it returns the right structure but with a different protocol version?

Comment (losfair) [reply test_util/src/lib.rs]: Done

File: cli/tests/unit/kv_test.ts [hunk 7]
```diff
@@ -1932,3 +1935,74 @@ Deno.test({
...
+    }
+  },
+});
+
+Deno.test({
+  name: "remote backend invalid format",
+  async fn() {
+    const db = await Deno.openKv(
+      "http://localhost:4545/kv_remote_authorize_invalid_format",
+    );
```
[the remark this answers is not in the document]
Comment (losfair) [reply cli/tests/unit/kv_test.ts:1959]: That slows down app startup - any user initialization logic after `Deno.openKv` would have to wait for the first KV metadata response.

I prefer to keep that an asynchronous operation and defer any error to the first KV API call.

**Review by ry — APPROVED** [review]
LGTM

there is some remaining concerns about introducing protoc as a build dependency - so we might need to come back and check-in the generated code.

**Comment by kitsonk** [comment]
Just want to comment that this threw me for a loop when I pulled main and did a compile. I suspect a lot of folks used to building from source will scratch their head as the cargo build error don't give any hints as to how to correct it and I actually had to read through the PR to figure out what to do.  Just has been years that `cargo build` was all you needed to do.

On the other hand, this is an awesome cool feature! 👍

**Comment by eternalphane** [comment]
> Just want to comment that this threw me for a loop when I pulled main and did a compile. I suspect a lot of folks used to building from source will scratch their head as the cargo build error don't give any hints as to how to correct it and I actually had to read through the PR to figure out what to do.

Maybe we should add a feature to switch the `remote` backend?

**Pull request merged**